ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,427 @@
+  zipLedgerTablesA f (TestTables x) (TestTables y) = TestTables <$> f x y
+  zipLedgerTables2A f (TestTables x) (TestTables y) (TestTables z) = TestTables <$> f x y z
+  foldLedgerTables f (TestTables x) = f x
+  foldLedgerTables2 f (TestTables x) (TestTables y) = f x y
+  namesLedgerTables = TestTables $ NameMK "TestTables"
+
+deriving instance Eq (LedgerTables TestLedger SeqDiffMK)
+
+data DbChangelogTestSetup = DbChangelogTestSetup
+  { operations          :: [Operation TestLedger]
+  , originalDbChangelog :: DbChangelog TestLedger
+  }
+
+data Operation l = Extend (l DiffMK) | Prune SecurityParam
+deriving instance Show (l DiffMK) => Show (Operation l)
+
+data DbChangelogTestSetupWithRollbacks = DbChangelogTestSetupWithRollbacks
+  { testSetup :: DbChangelogTestSetup
+  , rollbacks :: Int
+  } deriving (Show)
+
+instance Show DbChangelogTestSetup where
+  show = ppShow . operations
+
+instance Arbitrary DbChangelogTestSetup where
+
+  arbitrary = sized $ \n -> do
+    slotNo <- oneof [pure Origin, At <$> SlotNo <$> chooseEnum (1, 1000)]
+    operations <- genOperations slotNo n
+    pure $ DbChangelogTestSetup
+      { operations = operations
+      , originalDbChangelog = emptyDbChangelogAtSlot slotNo
+      }
+
+  shrink dblog = takeWhileJust $ tail (iterate reduce (Just dblog))
+    where
+      reduce (Just (DbChangelogTestSetup (_:ops) dblog')) = Just $ DbChangelogTestSetup ops dblog'
+      reduce _ = Nothing
+      takeWhileJust = catMaybes . takeWhile isJust
+
+instance Arbitrary DbChangelogTestSetupWithRollbacks where
+  arbitrary = do
+    setup <- arbitrary
+    let dblog = resultingDbChangelog setup
+    rollbacks <- chooseInt (0, AS.length (changelogVolatileStates dblog))
+    pure $ DbChangelogTestSetupWithRollbacks
+      { testSetup = setup
+      , rollbacks = rollbacks
+      }
+
+emptyDbChangelogAtSlot :: WithOrigin SlotNo -> DbChangelog TestLedger
+emptyDbChangelogAtSlot slotNo = emptyDbChangeLog (TestLedger ApplyEmptyMK $ pointAtSlot slotNo)
+
+resultingDbChangelog :: DbChangelogTestSetup -> DbChangelog TestLedger
+resultingDbChangelog setup = applyOperations (operations setup) (originalDbChangelog setup)
+
+applyOperations :: (TableStuff l, GetTip (l EmptyMK))
+  => [Operation l] -> DbChangelog l -> DbChangelog l
+applyOperations ops dblog = foldr' apply' dblog ops
+  where apply' (Extend newState) dblog' = extendDbChangelog dblog' newState
+        apply' (Prune sp) dblog'        = pruneVolatilePartDbChangelog sp dblog'
+
+
+{-------------------------------------------------------------------------------
+  Invariants
+-------------------------------------------------------------------------------}
+
+volatileTipAnchorsImmutable :: (GetTip (l EmptyMK), Eq (l EmptyMK)) => DbChangelog l -> Bool
+volatileTipAnchorsImmutable DbChangelog { changelogImmutableStates, changelogVolatileStates } =
+  AS.anchor changelogVolatileStates == AS.headAnchor changelogImmutableStates
```

Comment:
Would using `===` here give us more informative error messages should this property fail?